src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Cohost/OpenDocumentGenerator.cs

```diff
@@ -0,0 +1,182 @@
+
+        if (_documentManager.TryGetDocument(hostDocumentUri, out var documentSnapshot))
+        {
+            // Html
+            var htmlVirtualDocumentSnapshot = TryGetHtmlSnapshot(documentSnapshot);
+
+            // CSharp
+            var csharpVirtualDocumentSnapshot = await TryGetCSharpSnapshotAsync(documentSnapshot, document.Project.Key, version, cancellationToken).ConfigureAwait(false);
+
+            // Buffer work has to be on the UI thread
+            await _joinableTaskContext.Factory.SwitchToMainThreadAsync(cancellationToken);
+
+            Debug.Assert(htmlVirtualDocumentSnapshot is not null && csharpVirtualDocumentSnapshot is not null ||
+                htmlVirtualDocumentSnapshot is null && csharpVirtualDocumentSnapshot is null, "Found a Html XOR a C# document. Expected both or neither.");
+
+            if (htmlVirtualDocumentSnapshot is not null)
+            {
+                _documentManager.UpdateVirtualDocument<HtmlVirtualDocument>(
+                    hostDocumentUri,
+                    [new VisualStudioTextChange(0, htmlVirtualDocumentSnapshot.Snapshot.Length, generatedOutput.GetHtmlSourceText().ToString())],
+                    version,
+                    state: null);
+            }
+
+            if (csharpVirtualDocumentSnapshot is not null)
+            {
+                _documentManager.UpdateVirtualDocument<CSharpVirtualDocument>(
+                    hostDocumentUri,
+                    csharpVirtualDocumentSnapshot.Uri,
+                    [new VisualStudioTextChange(0, csharpVirtualDocumentSnapshot.Snapshot.Length, generatedOutput.GetCSharpSourceText().ToString())],
+                    version,
+                    state: null);
+                return;
+            }
+        }
+    }
+
+    private async Task<CSharpVirtualDocumentSnapshot?> TryGetCSharpSnapshotAsync(LSPDocumentSnapshot documentSnapshot, ProjectKey projectKey, int version, CancellationToken cancellationToken)
+    {
+        if (documentSnapshot.TryGetAllVirtualDocuments<CSharpVirtualDocumentSnapshot>(out var virtualDocuments))
+        {
+            if (virtualDocuments is [{ ProjectKey.Id: null }])
+            {
+                // If there is only a single virtual document, and its got a null id, then that means it's in our "misc files" project
+                // but the server clearly knows about it in a real project. That means its probably new, as Visual Studio opens a buffer
+                // for a document before we get the notifications about it being added to any projects. Lets try refreshing before
+                // we worry.
+                _logger.LogDebug("[Cohost] Refreshing virtual documents, and waiting for them, (for {hostDocumentUri})", documentSnapshot.Uri);
+
+                var task = _csharpVirtualDocumentAddListener.WaitForDocumentAddAsync(cancellationToken);
+                _documentManager.RefreshVirtualDocuments();
+                _ = await task.ConfigureAwait(true);
```

Comment:
I didn't see any threaded assertions about this code, is there a reason that we `ConfigureAwait(true)` here?

Reply:
Yep, thanks for catching this, this should have been run on the UI thread.